acpi/src/lib.rs

```diff
@@ -95,189 +100,173 @@ pub enum AcpiError {
     InvalidDsdtAddress,
     InvalidMadt(MadtError),
     InvalidGenericAddress,
+
+    AllocError,
 }
 
-pub struct AcpiTables<H>
-where
-    H: AcpiHandler,
-{
-    /// The revision of ACPI that the system uses, as inferred from the revision of the RSDT/XSDT.
-    pub revision: u8,
-    pub sdts: BTreeMap<sdt::Signature, Sdt>,
-    pub dsdt: Option<AmlTable>,
-    pub ssdts: Vec<AmlTable>,
+/// Common result type used by most result-returning functions.
+pub type AcpiResult<T> = core::result::Result<T, AcpiError>;
+
+/// All types representing ACPI tables should implement this trait.
+pub trait AcpiTable {
+    const SIGNATURE: Signature;
+
+    fn header(&self) -> &sdt::SdtHeader;
+
+    fn validate(&self) -> AcpiResult<()> {
+        self.header().validate(Self::SIGNATURE)
+    }
+}
+
+pub struct AcpiTables<H: AcpiHandler> {
+    mapping: PhysicalMapping<H, SdtHeader>,
+    revision: u8,
     handler: H,
 }
 
-impl<H> AcpiTables<H>
-where
-    H: AcpiHandler,
-{
+impl<H: AcpiHandler> AcpiTables<H> {
     /// Create an `AcpiTables` if you have the physical address of the RSDP.
-    pub unsafe fn from_rsdp(handler: H, rsdp_address: usize) -> Result<AcpiTables<H>, AcpiError> {
-        let rsdp_mapping = unsafe { handler.map_physical_region::<Rsdp>(rsdp_address, mem::size_of::<Rsdp>()) };
+    pub fn from_rsdp(handler: H, address: usize) -> AcpiResult<Self> {
```

Comment:
I'm not sure removing `unsafe` here is correct: people are still required to pass a valid pointer in to avoid UB no? (as in it we validate that it's a real RSDP or whatever, but it could still be problematic to dereference at all?

Reply:
I'm not sure—if they provide a bogus address, the validation will fail, yeah? But I suppose if we do also go ahead with optional checksumming, it could be an issue since validation won't be as reliable.

I'm for making it unsafe, in that case

Reply:
Yeah, but they could also provide an address that is e.g. non-canonical, or is in MMIO memory, or whatever right (i.e. even accessing it to validate it could be problematic). I'm generally in the mind that functions that accept arbitrary addresses and then dereference them in some way should be `unsafe`

Reply:
I'm convinced!